### DISCUSSION TITLE
AWS Bedrock Runtime - token count - requestId

### DISCUSSION THREAD

**Discussion by agustinp93** [aws/aws-sdk-js-v3#5879] [Q&A]
Hello!

My team is using `@aws-sdk/client-bedrock-runtime` and everything is working fine but we are missing metadata that is visible in other SDK's.

We are using: 
- AWS Lambda
- Node: version 20
- Bedrock: @aws-sdk/client-bedrock-runtime "^3.529.1"
-  modelId: "anthropic.claude-instant-v1", (any version we tried does not have the token count)

Request is based in this example: [https://docs.aws.amazon.com/bedrock/latest/userguide/bedrock-runtime_example_bedrock-runtime_InvokeClaude_section.html](url)

In the response we don't receive the token count of the request and response done to Bedrock:
`metadata {
  httpStatusCode: 200,
  requestId: '########-####-####-####-#########',
  extendedRequestId: undefined,
  cfId: undefined,
  attempts: 1,
  totalRetryDelay: 0
}`

I can see in the docs that other SDK like PHP do return headers in the response and the token count is one of them. I also saw them while doing requests from a React app.  But in this scenario where we moved to a lambda to be able to collect large amounts of data before invoking Bedrock this metadata we need is missing.

I'm missing something in the request config? is a bug we need to report? 

Thanks

**Comment by kimnamu** [comment]
Thanks for posting the full `$metadata` dump and the PHP comparison, that made this quick to reproduce.

Nothing is missing from your request config. The token counts come back as `x-amzn-bedrock-input-token-count` / `x-amzn-bedrock-output-token-count` response headers, and the Bedrock Runtime service model does not declare them for `InvokeModel`, so the SDK never maps them onto the output or `$metadata`. That matches the replies in #5917 and #8172 (the latter was closed as needing a change to the Bedrock service model).

Where the counts show up today (live run, `@aws-sdk/client-bedrock-runtime` 3.1140.0, Claude Haiku 4.5, us-east-1, same prompt each time):

| Call | Where | input / output |
|---|---|---|
| `InvokeModel` | `$metadata` | not there |
| `InvokeModel` | response headers (raw HTTP response only) | 9 / 5 |
| `InvokeModel`, Claude Messages body | `JSON.parse(body.transformToString()).usage` (`input_tokens` / `output_tokens`) | 9 / 5 |
| `Converse` | `usage.inputTokens` / `usage.outputTokens` (typed) | 9 / 5 |
| `InvokeModelWithResponseStream` | last chunk (`message_stop`), `amazon-bedrock-invocationMetrics` | 9 / 5 |

`Converse` is the only one of these that is part of the SDK's typed output, so it is the safest choice for metering (it is not in 3.529.1, so it needs a client upgrade). #5917 also notes the headers are not a modeled part of the API and could change. If you need to stay on `InvokeModel`, a middleware can copy them onto the output:

```js
import { BedrockRuntimeClient, InvokeModelCommand } from "@aws-sdk/client-bedrock-runtime";

const client = new BedrockRuntimeClient({ region: "us-east-1" });

// Add once, next to where the client is created (a second add with the same name throws).
client.middlewareStack.add(
  (next) => async (args) => {
    const result = await next(args); // call next() only once, a second call sends the request again
    const h = result.response.headers;
    const count = (name) => (name in h ? Number(h[name]) : undefined);
    result.output.tokenCounts = {
      input: count("x-amzn-bedrock-input-token-count"),
      output: count("x-amzn-bedrock-output-token-count"),
    };
    return result;
  },
  { step: "finalizeRequest", name: "bedrockTokenCounts" }
);

const { tokenCounts, $metadata } = await client.send(
  new InvokeModelCommand({
    modelId: "us.anthropic.claude-haiku-4-5-20251001-v1:0",
    contentType: "application/json",
    body: JSON.stringify({
      anthropic_version: "bedrock-2023-05-31",
      max_tokens: 5,
      messages: [{ role: "user", content: "Say hi" }],
    }),
  })
);
console.log($metadata.requestId, tokenCounts);
```

```
<request-id> { input: 9, output: 5 }
```

If you start from the middleware in #5917 instead, its first version returns `next(args)` a second time, which sends the request again. With a stub request handler, one `send()` reached the handler twice that way and once with the version above, on both 3.1140.0 and your 3.529.1.

<details>
<summary>Evidence (main 6050a38, Node 22)</summary>

- `InvokeModelResponse` has only `body`, `contentType`, `performanceConfigLatency`, `serviceTier` (`clients/client-bedrock-runtime/src/models/models_0.ts:5522-5546`), and no shape in `codegen/sdk-codegen/aws-models/bedrock-runtime.json` binds a token-count header. In 3.529.1 it was `body` and `contentType`.
- `ConverseResponse.usage: TokenUsage` is required (`models_0.ts:4432`, `TokenUsage` at `:4374`).

`InvokeModel` output with no middleware (3.1140.0):

```
[ 'body', 'contentType', '$metadata' ] [
  'httpStatusCode',
  'requestId',
  'extendedRequestId',
  'cfId',
  'attempts',
  'totalRetryDelay'
]
```

Raw headers and the other calls, same prompt (`max_tokens: 5`):

```
x-amzn-bedrock-* headers: {
  'x-amzn-bedrock-invocation-latency': '717',
  'x-amzn-bedrock-cache-write-input-token-count': '0',
  'x-amzn-bedrock-cache-read-input-token-count': '0',
  'x-amzn-bedrock-output-token-count': '5',
  'x-amzn-bedrock-input-token-count': '9'
}
body.usage              : {
  input_tokens: 9,
  cache_creation_input_tokens: 0,
  cache_read_input_tokens: 0,
  cache_creation: { ephemeral_5m_input_tokens: 0, ephemeral_1h_input_tokens: 0 },
  output_tokens: 5,
  service_tier: 'standard'
}
Converse usage          : {
  inputTokens: 9,
  outputTokens: 5,
  totalTokens: 14,
  cacheReadInputTokens: 0,
  cacheWriteInputTokens: 0
}
stream x-amzn-bedrock-* : { 'x-amzn-bedrock-content-type': 'application/json' }
stream last chunk      : {
  'amazon-bedrock-invocationMetrics': {
    inputTokenCount: 9,
    outputTokenCount: 5,
    invocationLatency: 638,
    firstByteLatency: 603
  }
}
```

Reading the body and the stream order (a plain `JSON.parse(body)` throws, the body is a `Uint8Array`):

```
JSON.parse(body) -> SyntaxError: Unexpected non-whitespace character after JSON at position 3
stream chunks: message_start > content_block_start > content_block_delta > content_block_delta > content_block_delta > content_block_stop > message_delta > message_stop+invocationMetrics
```

Request count with a stub `requestHandler` that counts calls (A = `return next(args)` after awaiting it, B = the middleware above):

```
A  next(args) twice -> requests sent: 2
B  next(args) once  -> requests sent: 1
   out.tokenCounts: { input: 12, output: 5 }
```

</details>
